### ISSUE TITLE
Consider faster clock mechanisms

### ISSUE THREAD

**Issue by MichaelXavier** [Soostone/katip#3]
We were advised that using formatTime heavily (especially in a tight loop) may be a bottleneck. This is used in the file handle scribe. We were also advised to consider using the `clock` package and using timespec for a more efficient timer. We've put a bit of consideration to this by allowing the user to specify a time function (which they can put behind an `auto-update` for slightly less accurate timestamps). I'm not sure how much I like this tradeoff because we still need to format UTC time in most of our scribes rather than unix epoch or some monotonic clock.

**Comment by ozataman** [comment]
Any chance the `thyme` lib has a faster formatter?

**Comment by MichaelXavier** [comment]
Looks like it if my criterion-foo is correct: https://gist.github.com/MichaelXavier/bba392149f2cb6f7510d

Mean for `time` formatting is 8.011 μs, mean for `thyme` is  `1.745 μs`, making `time` about 4x slower.

**Comment by mightybyte** [comment]
Just an FYI...right now the clock package is badly broken (see https://github.com/corsis/clock/issues/36), and I haven't received any response to that issue in almost three months.

**Comment by ozataman** [comment]
Wow, that seems potentially really bad. `thyme` seems like an easy speedup if we care about this.

**Comment by MichaelXavier** [comment]
The speedup would be easy to get but down side I see is that `time` is currently the incumbent and much more used package. Currently, the user's exposure to it is optional, but it comes in these instances:

1. If the user decides to override the autoupdate in LogEnv. We actually do this in some of our apps to get timestamps that are easier to linearize in elasticsearch.
2. If the user wants to provide an index sharding function with the soon to be released index sharding feature in katip-elasticsearch (this case should be far less frequent).

If the user needs to touch these settings, they'll have to include the `thyme` package in their direct dependencies and use the thyme equivalent.

**Comment by cetinsert** [comment]
`fromInteger` has been reintroduced in 0.7.1.1, package uploaded to hackage.

Now how could we address the clock issues [#32](https://github.com/corsis/clock/issues/32) and [#33](https://github.com/corsis/clock/issues/33)?

**Comment by sopvop** [comment]
For faster time formatting maybe it's better to just copy Aeson implementation from [here](https://github.com/bos/aeson/blob/master/Data/Aeson/Encoding/Builder.hs#L250) as it is really fast.

**Comment by MichaelXavier** [comment]
@sopvop great idea! The implementation looks simple enough and doesn't seem to incur any dependencies. I guess my only question is what the legalities are of using that algorithm. We already have an aeson dependency, so would that entitle us to reproduce the algorithm internally? I'd prefer that to depending on their internal module because its likely to change.

**Comment by mightybyte** [comment]
@MichaelXavier The aeson package is released under a BSD3 license, so I doubt copying the implementation would pose any legal problems.

**Comment by MichaelXavier** [comment]
I did a [quick experiment](https://github.com/Soostone/katip/compare/aeson-time-formatting?expand=1) where I ported in the aeson formatting code and set up a benchmark between the two implementations. Worth noting:

1. This benchmarks formatting a whole log line rather than just the timestamp portion. This seems like a good idea so we don't blow time formatting out of proportion with the amount of impact it will have.
2. Aeson's encoder targets a bytestring builder. Katip targets a text builder, so the current ported implementation coverts BS Builder -> LBS -> LT -> T Builder, so some efficiency is probably lost there.

Still, with these caveats, on my machine I'm getting a mean time of 23.23ns for the current formatter and 23.18ns for the naively ported aeson formatter. Its probably possible to get faster times by grokking the formatter that aeson has and trying to port it natively to text (though there seems to be way less low level exports in text builder so maybe this isn't possible), but with the benchmarks I'm getting, you cold expect 43 million line formats per second and I'm certain there's bigger perf fish to fry higher up the chain.
